File: src/llm/validation.test.ts
import assert from "node:assert";
import { describe, it } from "node:test";
import type { LLMContext } from "../analysis/types.js";
import {
  GLOBAL_BUILTINS,
  isValidIdentifier,
  RESERVED_WORDS,
  resolveConflict,
  sanitizeIdentifier,
  validateSuggestion
} from "./validation.js";

describe("validation", () => {
  describe("isValidIdentifier", () => {
    it("accepts valid identifiers", () => {
      assert.strictEqual(isValidIdentifier("foo"), true);
      assert.strictEqual(isValidIdentifier("_foo"), true);
      assert.strictEqual(isValidIdentifier("$foo"), true);
      assert.strictEqual(isValidIdentifier("foo123"), true);
      assert.strictEqual(isValidIdentifier("camelCase"), true);
      assert.strictEqual(isValidIdentifier("PascalCase"), true);
      assert.strictEqual(isValidIdentifier("_privateVar"), true);
      assert.strictEqual(isValidIdentifier("$jquery"), true);
    });

    it("rejects invalid identifiers", () => {
      assert.strictEqual(isValidIdentifier(""), false);
      assert.strictEqual(isValidIdentifier("123foo"), false);
      assert.strictEqual(isValidIdentifier("foo-bar"), false);
      assert.strictEqual(isValidIdentifier("foo bar"), false);
      assert.strictEqual(isValidIdentifier("foo.bar"), false);
    });
  });

  describe("sanitizeIdentifier", () => {
    it("removes invalid characters", () => {
      assert.strictEqual(sanitizeIdentifier("foo-bar"), "foobar");
      assert.strictEqual(sanitizeIdentifier("foo.bar"), "foobar");
      assert.strictEqual(sanitizeIdentifier("foo bar"), "foobar");
      assert.strictEqual(sanitizeIdentifier("foo@bar#baz"), "foobarbaz");
    });

    it("prefixes identifiers starting with numbers", () => {
      assert.strictEqual(sanitizeIdentifier("123foo"), "_123foo");
      assert.strictEqual(sanitizeIdentifier("1"), "_1");
    });

    it("handles empty input", () => {
      assert.strictEqual(sanitizeIdentifier(""), "_unnamed");
      // "@#$" becomes "$" because $ is a valid identifier character
      assert.strictEqual(sanitizeIdentifier("@#$"), "$");
      // Only truly empty result should give _unnamed
      assert.strictEqual(sanitizeIdentifier("@#%"), "_unnamed");
    });

    it("appends underscore to reserved words", () => {
      assert.strictEqual(sanitizeIdentifier("if"), "if_");
      assert.strictEqual(sanitizeIdentifier("class"), "class_");
      assert.strictEqual(sanitizeIdentifier("function"), "function_");
      assert.strictEqual(sanitizeIdentifier("return"), "return_");
    });

    it("appends underscore to global built-in names", () => {
      assert.strictEqual(sanitizeIdentifier("Date"), "Date_");
      assert.strictEqual(sanitizeIdentifier("Math"), "Math_");
      assert.strictEqual(sanitizeIdentifier("JSON"), "JSON_");
      assert.strictEqual(sanitizeIdentifier("console"), "console_");
      assert.strictEqual(sanitizeIdentifier("Promise"), "Promise_");
    });

    it("preserves valid identifiers", () => {
      assert.strictEqual(sanitizeIdentifier("validName"), "validName");
      assert.strictEqual(sanitizeIdentifier("_private"), "_private");
      assert.strictEqual(sanitizeIdentifier("$dollar"), "$dollar");
    });
  });

  describe("resolveConflict", () => {
    it("tries semantic suffixes first", () => {
      const used = new Set(["name"]);
      const result = resolveConflict("name", used);
      assert.strictEqual(result, "nameVal");
    });

    it("uses numeric suffix when semantic suffixes exhausted", () => {
      const used = new Set([
        "name",
        "nameVal",
        "nameVar",
        "nameRef",
        "nameItem",
        "nameData",
        "nameResult",
        "nameValue"
      ]);
      const result = resolveConflict("name", used);
      assert.strictEqual(result, "name2");
    });

    it("increments numeric suffix as needed", () => {
      const used = new Set([
        "name",
        "nameVal",
        "nameVar",
        "nameRef",
        "nameItem",
        "nameData",
        "nameResult",
        "nameValue",
        "name2",
        "name3"
      ]);
      const result = resolveConflict("name", used);
      assert.strictEqual(result, "name4");
    });

    it("falls back to underscore prefix as last resort", () => {
      // Create a set with all suffixes and numbers 2-100
      const used = new Set(["name"]);
      for (const suffix of [
        "Val",
        "Var",
        "Ref",
        "Item",
        "Data",
        "Result",
        "Value"
      ]) {
        used.add(`name${suffix}`);
      }
      for (let i = 2; i <= 100; i++) {
        used.add(`name${i}`);
      }

      const result = resolveConflict("name", used);
      assert.strictEqual(result, "_name");
    });

    it("never produces triple underscore prefixes", () => {
      // Exhaust all strategies before underscore to ensure no ___ stacking
      const used = new Set(["name"]);
      for (const suffix of [
        "Val",
        "Var",
        "Ref",
        "Item",
        "Data",
        "Result",
        "Value"
      ]) {
        used.add(`name${suffix}`);
      }
      for (let i = 2; i <= 100; i++) {
        used.add(`name${i}`);
      }
      used.add("_name");
      used.add("name_");
      used.add("local_name");
      used.add("inner_name");

      const result = resolveConflict("name", used);
      assert.ok(
        !result.includes("___"),
        `Result "${result}" should not contain ___`
      );
      assert.ok(
        !used.has(result),
        `Result "${result}" should not be in used set`
      );
    });

    it("uses trailing underscore when leading underscore is taken", () => {
      const used = new Set(["name"]);
      for (const suffix of [
        "Val",
        "Var",
        "Ref",
        "Item",
        "Data",
        "Result",
        "Value"
      ]) {
        used.add(`name${suffix}`);
      }
      for (let i = 2; i <= 100; i++) {
        used.add(`name${i}`);
      }
      used.add("_name");

      const result = resolveConflict("name", used);
      assert.strictEqual(result, "name_");
    });
  });

  describe("validateSuggestion", () => {
    const makeContext = (usedIdentifiers: string[] = []): LLMContext => ({
      functionCode: "function test() {}",
      calleeSignatures: [],
      callsites: [],
      usedIdentifiers: new Set(usedIdentifiers)
    });

    it("accepts valid suggestions", () => {
      const result = validateSuggestion({ name: "validName" }, makeContext());
      assert.strictEqual(result.valid, true);
    });

    it("rejects invalid identifier syntax", () => {
      const result = validateSuggestion({ name: "123invalid" }, makeContext());
      assert.strictEqual(result.valid, false);
      assert.ok(result.reason?.includes("Invalid"));
    });

    it("rejects reserved words", () => {
      const result = validateSuggestion({ name: "class" }, makeContext());
      assert.strictEqual(result.valid, false);
      assert.ok(result.reason?.includes("reserved"));
    });

    it("rejects global built-in names that would shadow globals", () => {
      for (const name of [
        "Date",
        "Math",
        "JSON",
        "Array",
        "Object",
        "Map",
        "Set",
        "Promise",
        "Error",
        "console",
        "File",
        "Blob",
        "globalThis"
      ]) {
        const result = validateSuggestion({ name }, makeContext());
        assert.strictEqual(result.valid, false, `Should reject "${name}"`);
        assert.ok(
          result.reason?.includes("global"),
          `Reason for "${name}" should mention global: ${result.reason}`
        );
      }
    });

    it("rejects names already in use", () => {
      const result = validateSuggestion(
        { name: "existingVar" },
        makeContext(["existingVar", "otherVar"])
      );
      assert.strictEqual(result.valid, false);
      assert.ok(result.reason?.includes("already in use"));
    });

    it("rejects overly long names", () => {
      const longName = "a".repeat(51);
      const result = validateSuggestion({ name: longName }, makeContext());
      assert.strictEqual(result.valid, false);
      assert.ok(result.reason?.includes("too long"));
    });

    it("accepts names at the length limit", () => {
      const maxName = "a".repeat(50);
      const result = validateSuggestion({ name: maxName }, makeContext());
      assert.strictEqual(result.valid, true);
    });
  });

  describe("RESERVED_WORDS", () => {
    it("contains common JavaScript keywords", () => {
      const expected = [
        "if",
        "else",
        "for",
        "while",
        "function",
        "class",
        "const",
        "let",
        "var",
        "return",
        "await"
        // Note: "async" is not a reserved word, only "await" is
      ];
      for (const word of expected) {
        assert.ok(RESERVED_WORDS.has(word), `Should contain "${word}"`);
      }
    });

    it("contains literal values", () => {
      assert.ok(RESERVED_WORDS.has("null"));
      assert.ok(RESERVED_WORDS.has("true"));
      assert.ok(RESERVED_WORDS.has("false"));
      assert.ok(RESERVED_WORDS.has("undefined"));
    });
  });

  describe("GLOBAL_BUILTINS", () => {
    it("contains core JavaScript globals", () => {
      const expected = [
        "Date",
        "Math",
        "JSON",
        "Array",
        "Object",
        "Map",
        "Set",
        "Promise",
        "Error",
        "RegExp",
        "Symbol",
        "Proxy",
        "Reflect",
        "console",
        "globalThis"
      ];
      for (const name of expected) {
        assert.ok(GLOBAL_BUILTINS.has(name), `Should contain "${name}"`);
      }
    });

    it("contains Web/Node.js platform globals", () => {
      const expected = [
        "File",
        "Blob",
        "FormData",
        "URL",
        "URLSearchParams",
        "Buffer",
        "process",
        "setTimeout",
        "setInterval",
        "clearTimeout",
        "clearInterval",
        "AbortController"
      ];
      for (const name of expected) {
        assert.ok(GLOBAL_BUILTINS.has(name), `Should contain "${name}"`);
      }
    });
  });
});

File: crates/humanify-core/src/naming/validation/validation_test.rs
//! Ported fixture-for-fixture from src/llm/validation.test.ts, plus the
//! probe vectors (test/parity/wp42-vectors.json, recorded from the real TS).

use std::collections::HashSet;

use super::*;
use crate::naming::test_vectors;

fn used(names: &[&str]) -> HashSet<String> {
    names.iter().map(|s| s.to_string()).collect()
}

fn conflict(name: &str, set: &HashSet<String>) -> String {
    resolve_conflict(name, |c| set.contains(c))
}

#[test]
fn accepts_valid_identifiers() {
    for name in [
        "foo",
        "_foo",
        "$foo",
        "foo123",
        "camelCase",
        "PascalCase",
        "_privateVar",
        "$jquery",
    ] {
        assert!(is_valid_identifier(name), "{name}");
    }
}

#[test]
fn rejects_invalid_identifiers() {
    for name in ["", "123foo", "foo-bar", "foo bar", "foo.bar"] {
        assert!(!is_valid_identifier(name), "{name}");
    }
}

#[test]
fn sanitize_removes_invalid_characters() {
    assert_eq!(sanitize_identifier("foo-bar"), "foobar");
    assert_eq!(sanitize_identifier("foo.bar"), "foobar");
    assert_eq!(sanitize_identifier("foo bar"), "foobar");
    assert_eq!(sanitize_identifier("foo@bar#baz"), "foobarbaz");
}

#[test]
fn sanitize_prefixes_identifiers_starting_with_numbers() {
    assert_eq!(sanitize_identifier("123foo"), "_123foo");
    assert_eq!(sanitize_identifier("1"), "_1");
}

#[test]
fn sanitize_handles_empty_input() {
    assert_eq!(sanitize_identifier(""), "_unnamed");
    // "@#$" strips to "$", which is a global builtin (jQuery) → suffixed
    assert_eq!(sanitize_identifier("@#$"), "$_");
    assert_eq!(sanitize_identifier("@#%"), "_unnamed");
}

#[test]
fn sanitize_appends_underscore_to_reserved_words() {
    assert_eq!(sanitize_identifier("if"), "if_");
    assert_eq!(sanitize_identifier("class"), "class_");
    assert_eq!(sanitize_identifier("function"), "function_");
    assert_eq!(sanitize_identifier("return"), "return_");
}

#[test]
fn sanitize_appends_underscore_to_global_builtins() {
    for name in ["Date", "Math", "JSON", "console", "Promise"] {
        assert_eq!(sanitize_identifier(name), format!("{name}_"));
    }
}

#[test]
fn sanitize_preserves_valid_identifiers() {
    for name in ["validName", "_private", "$dollar"] {
        assert_eq!(sanitize_identifier(name), name);
    }
}

#[test]
fn resolve_conflict_tries_semantic_suffixes_first() {
    assert_eq!(conflict("name", &used(&["name"])), "nameVal");
}

#[test]
fn resolve_conflict_uses_numeric_suffix_when_semantic_exhausted() {
    let set = used(&[
        "name",
        "nameVal",
        "nameVar",
        "nameRef",
        "nameItem",
        "nameData",
        "nameResult",
        "nameValue",
    ]);
    assert_eq!(conflict("name", &set), "name2");
}

#[test]
fn resolve_conflict_increments_numeric_suffix() {
    let set = used(&[
        "name",
        "nameVal",
        "nameVar",
        "nameRef",
        "nameItem",
        "nameData",
        "nameResult",
        "nameValue",
        "name2",
        "name3",
    ]);
    assert_eq!(conflict("name", &set), "name4");
}

#[test]
fn resolve_conflict_stays_numeric_past_100() {
    let mut set = used(&["name"]);
    for suffix in DECORATION_WORDS {
        set.insert(format!("name{suffix}"));
    }
    for i in 2..=100 {
        set.insert(format!("name{i}"));
    }
    assert_eq!(conflict("name", &set), "name101");
}

/// The TS test walks 1,200 rungs asserting stem recovery through
/// prior-name-snap's `nameStem` (WP4.5, not yet ported). The parity form is
/// stronger: the whole ladder, rung for rung, equals the TS's — through the
/// 999 numeric rungs into the `nameVal2…` decorated-numeric tail.
#[test]
fn conflict_ladders_match_the_ts_rung_for_rung() {
    let v = test_vectors();
    for (name, ladder) in v["conflictLadders"].as_object().unwrap() {
        let mut set = used(&[name.as_str()]);
        for (i, expected) in ladder.as_array().unwrap().iter().enumerate() {
            let got = conflict(name, &set);
            assert_eq!(got, expected.as_str().unwrap(), "{name} rung {i}");
            assert!(set.insert(got));
        }
    }
}

#[test]
fn reserved_words_contain_keywords_literals_and_arguments() {
    for w in [
        "if",
        "else",
        "for",
        "while",
        "function",
        "class",
        "const",
        "let",
        "var",
        "return",
        "await",
        "null",
        "true",
        "false",
        "undefined",
        "arguments",
    ] {
        assert!(is_reserved_word(w), "{w}");
    }
    // "async" is not a reserved word, only "await" is
    assert!(!is_reserved_word("async"));
}

#[test]
fn global_builtins_contain_core_platform_and_host_globals() {
    for name in [
        "Date",
        "Math",
        "JSON",
        "Array",
        "Object",
        "Map",
        "Set",
        "Promise",
        "Error",
        "RegExp",
        "Symbol",
        "Proxy",
        "Reflect",
        "console",
        "globalThis",
        "File",
        "Blob",
        "FormData",
        "URL",
        "URLSearchParams",
        "Buffer",
        "process",
        "setTimeout",
        "setInterval",
        "clearTimeout",
        "clearInterval",
        "AbortController",
        "window",
        "document",
        "self",
        "location",
        "navigator",
        "$",
        "jQuery",
        "define",
        "Bun",
        "importScripts",
        "postMessage",
    ] {
        assert!(is_global_builtin(name), "{name}");
    }
}

#[test]
fn global_builtins_do_not_blanket_forbid_browser_names() {
    for name in ["event", "status", "name", "history", "screen"] {
        assert!(!is_global_builtin(name), "{name}");
    }
}

/// The three tables equal the TS's, in the TS's order (GLOBAL_BUILTINS is
/// generated from the pinned `globals` package — a bump shows up here).
#[test]
fn tables_equal_the_ts_tables() {
    let v = test_vectors();
    let list = |key: &str| -> Vec<String> {
        v["tables"][key]
            .as_array()
            .unwrap()
            .iter()
            .map(|s| s.as_str().unwrap().to_string())
            .collect()
    };
    assert_eq!(GLOBAL_BUILTINS.to_vec(), list("globalBuiltins"));
    assert_eq!(RESERVED_WORDS.to_vec(), list("reservedWords"));
    assert_eq!(DECORATION_WORDS.to_vec(), list("decorationWords"));
}

#[test]
fn validation_vectors_match_the_ts() {
    let v = test_vectors();
    for row in v["validation"].as_array().unwrap() {
        let input = row["input"].as_str().unwrap();
        assert_eq!(
            is_valid_identifier(input),
            row["isValid"].as_bool().unwrap(),
            "isValid {input:?}"
        );
        assert_eq!(
            sanitize_identifier(input),
            row["sanitized"].as_str().unwrap(),
            "sanitize {input:?}"
        );
    }
}
